File: python/tournament/tournament.py
class Table:
    def __init__(self):
        self._teams = {}

    def add_match(self, t1_name, t2_name, result):
        t1 = Team(t1_name)
        if t1_name in self._teams:
            t1 = self._teams[t1_name]

        t2 = Team(t2_name)
        if t2_name in self._teams:
            t2 = self._teams[t2_name]

        if result == "draw":
            t1.add_draw()
            t2.add_draw()
        elif result == "win":
            t1.add_win()
            t2.add_loss()
        elif result == "loss":
            t2.add_win()
            t1.add_loss()
        else:
            raise Exception(f"unrecognized result: {result}")

        self._teams[t1_name], self._teams[t2_name] = t1, t2

    def render(self):
        header = f"{'Team': <30}"
        header += f" | {'MP': >2}"
        header += f" | {'W': >2}"
        header += f" | {'D': >2}"
        header += f" | {'L': >2}"
        header += f" | {'P': >2}"

        results = [
            header,
        ]
        for team in sorted(self._teams.values(), reverse=True):
            row = f"{str(team): <30}"
            row += f" | {team.matches_played: >2}"
            row += f" | {team.wins: >2}"
            row += f" | {team.draws: >2}"
            row += f" | {team.losses: >2}"
            row += f" | {team.points: >2}"
            results.append(row)

        return results


class Team:
    results_to_points = {"win": 3, "draw": 1, "loss": 0}

    def __init__(self, name):
        self.name = name
        self.matches_played = 0
        self.wins = 0
        self.losses = 0
        self.draws = 0

    @property
    def points(self):
        pts = Team.results_to_points["win"] * self.wins
        pts += Team.results_to_points["draw"] * self.draws
        pts += Team.results_to_points["loss"] * self.losses

        return pts

    def add_win(self):
        self.matches_played += 1
        self.wins += 1

    def add_draw(self):
        self.matches_played += 1
        self.draws += 1

    def add_loss(self):
        self.matches_played += 1
        self.losses += 1

    def __repr__(self):
        return str(self.name)

    def __hash__(self):
        return hash(repr(self))

    def __lt__(self, other):
        a = (-1 * self.points, self.matches_played, self.name)
        b = (-1 * other.points, other.matches_played, other.name)

        return b < a


def tally(matches):
    table = Table()
    for match in matches:
        table.add_match(*match.split(";"))

    return table.render()


def main():
    table = tally(
        [
            "Courageous Californians;Devastating Donkeys;win",
            "Allegoric Alaskans;Blithering Badgers;win",
            "Devastating Donkeys;Allegoric Alaskans;loss",
            "Courageous Californians;Blithering Badgers;win",
            "Blithering Badgers;Devastating Donkeys;draw",
            "Allegoric Alaskans;Courageous Californians;draw",
        ]
    )
    for row in table:
        print(row)


if __name__ == "__main__":
    main()

File: go/tournament/tournament.go
// Package tournament implements a solution for the exercism tournament challenge
package tournament

import (
	"encoding/csv"
	"fmt"
	"io"
	"sort"
)

// Team represents a team in a footbal competition
type Team struct {
	name          string
	matchesPlayed int
	wins          int
	draws         int
	losses        int
}

// Table represents a table of results of matches played between teams
type Table struct {
	teams map[string]*Team
}

func (team *Team) points() int {
	return 3*team.wins + team.draws
}

func (team *Team) win() {
	team.matchesPlayed++
	team.wins++
}

func (team *Team) loose() {
	team.matchesPlayed++
	team.losses++
}

func (team *Team) draw() {
	team.matchesPlayed++
	team.draws++
}

func (t *Table) addMatch(team1Name, team2Name, result string) error {
	var team1, team2 *Team
	if team, ok := t.teams[team1Name]; ok {
		team1 = team
	} else {
		team1 = &Team{name: team1Name}
	}
	if team, ok := t.teams[team2Name]; ok {
		team2 = team
	} else {
		team2 = &Team{name: team2Name}
	}
	if result == "draw" {
		team1.draw()
		team2.draw()
	} else if result == "win" {
		team1.win()
		team2.loose()
	} else if result == "loss" {
		team1.loose()
		team2.win()
	} else {
		return fmt.Errorf("invalid result '%v'", result)
	}
	t.teams[team1Name] = team1
	t.teams[team2Name] = team2
	return nil
}

func (t *Table) getSortedTeams() []*Team {
	teams := make([]*Team, 0, len(t.teams))
	for _, v := range t.teams {
		teams = append(teams, v)
	}
	sort.Slice(teams, func(p, q int) bool {
		if teams[p].points() != teams[q].points() {
			return teams[p].points() > teams[q].points()
		}
		return teams[p].name < teams[q].name
	})
	return teams
}

// Tally reads a stream of matches and writes a stream representing the table
func Tally(i io.Reader, o io.Writer) error {
	table := &Table{teams: map[string]*Team{}}
	r := csv.NewReader(i)
	r.Comma = ';'
	r.Comment = '#'
	r.FieldsPerRecord = 3
	for {
		record, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return err
		}
		err = table.addMatch(record[0], record[1], record[2])
		if err != nil {
			return err
		}
	}
	l := "%-31v|%3v |%3v |%3v |%3v |%3v\n"
	fmt.Fprintf(o, l, "Team", "MP", "W", "D", "L", "P")
	for _, t := range table.getSortedTeams() {
		fmt.Fprintf(o, l, t.name, t.matchesPlayed, t.wins, t.draws, t.losses, t.points())
	}
	return nil
}
